internal/cachecontroller/cache_controller.go

```diff
@@ -152,13 +153,15 @@ func (c *InMemoryCacheController) findChangesAndInvalidate(ctx context.Context,
 		LastModified: changes[0].GetTimestamp().AsTime(),
 	}
 
+	hasCacheEntry := c.cache.Get(cacheKey) != nil
+
 	// set changelog entry as soon as possible for subsequent cache
 	// lookups have the entry and not have to wait on the existing singleflight group
 	c.cache.Set(cacheKey, entry, c.ttl)
 
 	lastVerified := time.Now().Add(-c.ttl)
 
-	if entry.LastModified.Before(lastVerified) {
+	if hasCacheEntry && entry.LastModified.Before(lastVerified) {
```

Comment:
> The algorithm needs to take care of the case where the lastVerified time was not set in the cache (either expired or first time).

I think line 164 relates to this item in your PR description but I don't understand it. Why do we care if the cacheKey existed before? We are overwriting it line 160 anyway. I'm confused

Reply:
Imagine that in the case of findChangesAndInvalidate is ran for the first time (which is found via entry not present), we will need to invalidate the iterator cache because we do not know whether the iterator cache is set before / after the last read changes.

Reply:
This took me a while to understand, longer than i'd like to admit. I got thrown off by `hasCacheEntry` because i thought it was related to the `entry` variable, but it's not related at all. Can we rename?

- `hasCacheEntry` -> `lastInvalidationOccurred`? 
-  `lastVerified` -> `timestampOfLastInvalidation`?

Also, i put some debug logs and i cannot see a test where we exercise the `false && false` scenario.

Reply:
updated as noted above

Reply:
Much appreciated :)